subsys/fs/nvs/nvs.c

```diff
@@ -1198,14 +1250,38 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, uint16_t id, void *data, size_t len,
 		return -ENOENT;
 	}
 
+#ifdef CONFIG_NVS_DATA_CRC
+	/* When data CRC is enabled, there should be at least the CRC stored in the data field */
+	if (wlk_ate.len < NVS_DATA_CRC_SIZE) {
+		return -ENOENT;
+	}
+#endif
+
 	rd_addr &= ADDR_SECT_MASK;
 	rd_addr += wlk_ate.offset;
-	rc = nvs_flash_rd(fs, rd_addr, data, MIN(len, wlk_ate.len));
+	rc = nvs_flash_rd(fs, rd_addr, data, MIN(len, wlk_ate.len - NVS_DATA_CRC_SIZE));
 	if (rc) {
 		goto err;
 	}
 
-	return wlk_ate.len;
+	/* Check data CRC */
+#ifdef CONFIG_NVS_DATA_CRC
+	rd_addr += wlk_ate.len - NVS_DATA_CRC_SIZE;
+	rc = nvs_flash_rd(fs, rd_addr, &read_data_crc, sizeof(read_data_crc));
+	if (rc) {
+		goto err;
+	}
+
+	computed_data_crc = crc32_ieee(data, wlk_ate.len - NVS_DATA_CRC_SIZE);
```

Comment:
This calculation will fail if `len < wlk_ate.len`. The data buffer size can be smaller then `ate.len`.

Reply:
You're correct ! Are you OK to disable CRC computation for partial reads ? CRC computation will be checked only for complete reads, which makes sense because if the CRC was implemented in a Zephyr application, the application would have need to read entirely the data to retrieve the CRC stored at the end.
I can update the documentation accordingly.

Reply:
> You're correct ! Are you OK to disable CRC computation for partial reads ? CRC computation will be checked only for complete reads, which makes sense because if the CRC was implemented in a Zephyr application, the application would have need to read entirely the data to retrieve the CRC stored at the end. I can update the documentation accordingly.

Yes, I am OK with that.